rust/src/daemon.rs

```diff
@@ -131,6 +149,167 @@ fn deployment_populate_variant_origin(
+            }
+            match e {
+                Ok(()) => SocketMessage::ServerOk,
+                Err(e) => SocketMessage::ServerError {
+                    msg: format!("{e}"),
+                },
+            }
+        }
+        o => SocketMessage::ServerError {
+            msg: format!("Unexpected message: {o:?}"),
+        },
+    };
+    write_message(&client, reply).context("Writing client reply")?;
+    tracing::debug!("Acknowleged client");
+    Ok(())
+}
+
+static SHUTDOWN_SIGNAL: Lazy<Mutex<Option<Sender<()>>>> = Lazy::new(|| Mutex::new(None));
+
+fn run_acknowledgement_worker(listener: OwnedSocketlike) {
+    tracing::debug!("Processing clients...");
+    loop {
+        let sock = match rustix::net::accept(&listener) {
+            Ok(s) => s,
+            Err(e) => {
+                tracing::warn!("Failed to accept client: {e}");
+                continue;
+            }
+        };
+        std::thread::spawn(move || {
+            if let Err(e) = client_hello(sock.into_socketlike(), Ok(())) {
+                tracing::warn!("error acknowledging client: {e}");
+            }
+        });
+    }
+}
+
+/// Ensure all asynchronous tasks in this Rust half of the daemon code are stopped.
+/// Called from C++.
+pub(crate) fn daemon_terminate() {
+    if let Some(chan) = (*SHUTDOWN_SIGNAL).lock().unwrap().take() {
+        let _ = chan.send(());
+    }
+}
+
+fn process_one_client(listener: OwnedSocketlike, e: anyhow::Error) -> Result<()> {
+    let incoming = rustix::net::accept(&listener)?;
+    // Send the error to the client
+    client_hello(incoming.into_socketlike(), Err(e))?;
+    // We've successfully sent the error; the caller of this function will
+    // propagate the error.
+    Ok(())
+}
+
+/// Perform initialization steps required by systemd service activation.
+///
+/// This ensures that the system is running under systemd, then receives the
+/// socket-FD for main IPC logic, and notifies systemd about ready-state.
+pub(crate) fn daemon_main(debug: bool) -> Result<()> {
+    let handle = tokio::runtime::Handle::current();
+    let _tokio_guard = handle.enter();
+    if !systemd::daemon::booted()? {
+        return Err(anyhow!("not running as a systemd service"));
+    }
+
+    let init_res: Result<()> = crate::ffi::daemon_init_inner(debug).map_err(|e| e.into());
+    tracing::debug!("Initialization result: {init_res:?}");
+
+    let mut fds = systemd::daemon::listen_fds(false)?.iter();
+    let (listener, init_res) = match fds.next() {
+        None => {
+            // If started directly via `systemctl start` or DBus activation, we
+            // directly propagate the error back to our exit code without even bothering
+            // with a socket.
+            init_res.context("Initializing via dbus")?;
+            tracing::debug!("Initializing directly (not socket activated)");
+            let listener = cfg!(feature = "client-socket")
```

Comment:
The `None` case would not exist if we can no longer start the daemon from dbus and only activated via the socket

Reply:
Ahh...this is a complicated topic!  We definitely must continue to support clients that aren't `rpm-ostree` starting the daemon via DBus.

So hmm...two ways to do that.

One is just having an integration test that's like
```
systemctl stop rpm-ostreed
rpmostree_busctl_call_os ...
```

Reply:
Or could it?

Reply:
yes this fails since when the daemon is `idle` or `stopped`, and `systemctl start` or `busctl` is invoked. We may need a more graceful handling of re-connecting to the socket.